### PR TITLE
close #33 直進する

### LINKED ISSUE #33: 直進する
## 完了条件
距離で直進するクラスと色まで直進するクラスの共通部分を抜き出して基底クラスとすること。
※できているかどうか単体でテストできないのでレビュー頼りになります。

## 実現方法
- 直進するクラスの共通部分を抜き出す。
- 実装していない関数に関しては完全仮想関数として実装して必ず派生クラス上でオーバーライドするようにする。

## 関連資料（省略可）
[DistanceStraight.cpp](https://github.com/KatLab-MiyazakiUniv/etrobocon2022/blob/main/module/Motion/DistanceStraight.cpp)
[DistanceStraight.h](https://github.com/KatLab-MiyazakiUniv/etrobocon2022/blob/main/module/Motion/DistanceStraight.h)
[ColorStraight.cpp](https://github.com/KatLab-MiyazakiUniv/etrobocon2022/blob/main/module/Motion/ColorStraight.h)
[ColorStraight.h](https://github.com/KatLab-MiyazakiUniv/etrobocon2022/blob/main/module/Motion/ColorStraight.h)

### PR DESCRIPTION
**Pull request by Negimarge**
# チェックリスト

- [x] clang-format している
- [x] コーディング規約に準じている
- [x] チケットの完了条件を満たしている

# 変更点
 - 直進の中間クラスを追加

### CODE REVIEW

**Comment by Negimarge** [comment]
# make buildした時のログ
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ make build
cd /home/nakiz/etrobo && make app=etrobocon2023
invoke: make app=etrobocon2023
make[1]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/workspace'
rm -rf /home/nakiz/etrobo/hrp3/sdk/workspace/.././OBJ/
Generating Makefile from ../common/Makefile.app.
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
rm -f cfg1_out cfg1_out.o cfg1_out.c cfg1_out.syms cfg1_out.srec module_cfg.h module_cfg.c \#* *~ *.o
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
  CFG[1]  module_cfg.h
[cfg.rb] Generated cfg1_out.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CFG[2]  module_cfg.h
[cfg.rb] Generated module_cfg.c
[cfg.rb] Generated module_cfg.h
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
  CC      module_cfg.c
  CC      ../../library/t_perror.c
  CC      ../../library/strerror.c
  CC      ../../library/vasyslog.c
  CC      ../../target/ev3_gcc/TLSF-2.4.6/src/tlsf.c
  CC      ../common/ev3api/src/ev3api.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_battery.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_brick.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_fs.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_speaker.c
  CC      ../common/ev3api/src/ev3api_lcd.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_motor.c
  CC      ../common/ev3api/src/ev3api_newlib.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_sensor.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calibrator.cpp
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/EtRobocon2023.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Controller.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Measurer.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Timer.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/ColorJudge.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/Mileage.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/Pid.cpp
make[2]: *** No rule to make target 'Logger.cpp', needed by 'Logger.o'.  Stop.
make[2]: *** Waiting for unfinished jobs....
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp
In file included from /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp:6:0:
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h: In constructor 'SpeedCalculator::SpeedCalculator(double)':
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h:33:16: warning: 'SpeedCalculator::targetSpeed' will be initialized after [-Wreorder]
   const double targetSpeed;
                ^~~~~~~~~~~
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h:31:7: warning:   'Pid SpeedCalculator::pid' [-Wreorder]
   Pid pid;
       ^~~
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp:8:1: warning:   when initialized here [-Wreorder]
 SpeedCalculator::SpeedCalculator(double _targetSpeed)
 ^~~~~~~~~~~~~~~
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[1]: *** [../common/Makefile.workspace:97: app] Error 2
make[1]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/workspace'
fakemake on HRP3: make failed ... app=etrobocon2023
make: *** [Makefile:24: build] Error 1
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git branch
  main
* ticket-33
  work-33
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ make build
cd /home/nakiz/etrobo && make app=etrobocon2023
invoke: make app=etrobocon2023
make[1]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/workspace'
rm -rf /home/nakiz/etrobo/hrp3/sdk/workspace/.././OBJ/
Generating Makefile from ../common/Makefile.app.
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
rm -f cfg1_out cfg1_out.o cfg1_out.c cfg1_out.syms cfg1_out.srec module_cfg.h module_cfg.c \#* *~ *.o
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
  CFG[1]  module_cfg.h
[cfg.rb] Generated cfg1_out.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CFG[2]  module_cfg.h
[cfg.rb] Generated module_cfg.c
[cfg.rb] Generated module_cfg.h
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
  CC      module_cfg.c
  CC      ../../library/t_perror.c
  CC      ../../library/strerror.c
  CC      ../../library/vasyslog.c
  CC      ../../target/ev3_gcc/TLSF-2.4.6/src/tlsf.c
  CC      ../common/ev3api/src/ev3api.c
  CC      ../common/ev3api/src/ev3api_battery.c
  CC      ../common/ev3api/src/ev3api_brick.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_fs.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_speaker.c
  CC      ../common/ev3api/src/ev3api_lcd.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_motor.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_newlib.c
  CC      ../common/ev3api/src/ev3api_sensor.c
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calibrator.cpp
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/EtRobocon2023.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Controller.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Measurer.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Timer.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/ColorJudge.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/Mileage.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/Pid.cpp
make[2]: *** No rule to make target 'Logger.cpp', needed by 'Logger.o'.  Stop.
make[2]: *** Waiting for unfinished jobs....
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp
In file included from /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp:6:0:
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h: In constructor 'SpeedCalculator::SpeedCalculator(double)':
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h:33:16: warning: 'SpeedCalculator::targetSpeed' will be initialized after [-Wreorder]
   const double targetSpeed;
                ^~~~~~~~~~~
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h:31:7: warning:   'Pid SpeedCalculator::pid' [-Wreorder]
   Pid pid;
       ^~~
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp:8:1: warning:   when initialized here [-Wreorder]
 SpeedCalculator::SpeedCalculator(double _targetSpeed)
 ^~~~~~~~~~~~~~~
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[1]: *** [../common/Makefile.workspace:97: app] Error 2
make[1]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/workspace'
fakemake on HRP3: make failed ... app=etrobocon2023
make: *** [Makefile:24: build] Error 1
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git sta
stage    stash    status   
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git sta
stage    stash    status   
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git status
On branch ticket-33
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   module/Motion/Straight.cpp

no changes added to commit (use "git add" and/or "git commit -a")
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git add module/Motion/Straight.cpp 
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git commit -m "fix: pwd, currentPwdの型をint型に"
[ticket-33 c88c113] fix: pwd, currentPwdの型をint型に
 1 file changed, 2 insertions(+), 2 deletions(-)
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ git push origin ticket-33
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 8 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 541 bytes | 180.00 KiB/s, done.
Total 5 (delta 3), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To https://github.com/KatLab-MiyazakiUniv/etrobocon2023.git
   20f24ce..c88c113  ticket-33 -> ticket-33
nakiz@LAPTOP-MUDS86N5:~/etrobo/workspace/etrobocon2023$ make build
cd /home/nakiz/etrobo && make app=etrobocon2023
invoke: make app=etrobocon2023
make[1]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/workspace'
rm -rf /home/nakiz/etrobo/hrp3/sdk/workspace/.././OBJ/
Generating Makefile from ../common/Makefile.app.
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
rm -f cfg1_out cfg1_out.o cfg1_out.c cfg1_out.syms cfg1_out.srec module_cfg.h module_cfg.c \#* *~ *.o
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
  CFG[1]  module_cfg.h
[cfg.rb] Generated cfg1_out.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CFG[2]  module_cfg.h
[cfg.rb] Generated module_cfg.c
[cfg.rb] Generated module_cfg.h
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[2]: Entering directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
  CC      module_cfg.c
  CC      ../../library/t_perror.c
  CC      ../../library/strerror.c
  CC      ../../library/vasyslog.c
  CC      ../../target/ev3_gcc/TLSF-2.4.6/src/tlsf.c
  CC      ../common/ev3api/src/ev3api.c
  CC      ../common/ev3api/src/ev3api_battery.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_brick.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_fs.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_speaker.c
  CC      ../common/ev3api/src/ev3api_lcd.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CC      ../common/ev3api/src/ev3api_newlib.c
  CC      ../common/ev3api/src/ev3api_motor.c
  CC      ../common/ev3api/src/ev3api_sensor.c
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calibrator.cpp
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
cc1: warning: command line option '-std=gnu++11' is valid for C++/ObjC++ but not for C
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/EtRobocon2023.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Controller.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Measurer.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/API/Timer.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/ColorJudge.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/Mileage.cpp
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/Pid.cpp
make[2]: *** No rule to make target 'Logger.cpp', needed by 'Logger.o'.  Stop.
make[2]: *** Waiting for unfinished jobs....
  CXX     /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp
In file included from /home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp:6:0:
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h: In constructor 'SpeedCalculator::SpeedCalculator(double)':
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h:33:16: warning: 'SpeedCalculator::targetSpeed' will be initialized after [-Wreorder]
   const double targetSpeed;
                ^~~~~~~~~~~
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.h:31:7: warning:   'Pid SpeedCalculator::pid' [-Wreorder]
   Pid pid;
       ^~~
/home/nakiz/etrobo/hrp3/sdk/workspace/etrobocon2023/module/Calculator/SpeedCalculator.cpp:8:1: warning:   when initialized here [-Wreorder]
 SpeedCalculator::SpeedCalculator(double _targetSpeed)
 ^~~~~~~~~~~~~~~
make[2]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/OBJ'
make[1]: *** [../common/Makefile.workspace:97: app] Error 2
make[1]: Leaving directory '/home/nakiz/etrobo/hrp3/sdk/workspace'
fakemake on HRP3: make failed ... app=etrobocon2023
make: *** [Makefile:24: build] Error 1

**Comment by Negimarge** [comment]
#make build時のスクリーンショット
スクリーンショット (7)
スクリーンショット (8)
スクリーンショット (9)

File: module/Motion/Straight.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    // PWM値を目標速度値に合わせる
+    SpeedCalculator SpeedCalculator(targetSpeed);
+    currentPwm = SpeedCalculator.calcPwmFromSpeed();
+
+    // 左右のモーターカウントを合わせるための補正値計算
+    error = (currentLeftMotorCount - initialLeftMotorCount)
+            - (currentRightMotorCount - initialRightMotorCount);
+    adjustmentPwm = static_cast<int>(pid.calculatePid(error));
```
Comment (miyashita64) [inline module/Motion/Straight.cpp]: 左右の誤差等、この辺りのどの動作にも共通する誤差は、SpeedCalculator辺りが調整べきことな気がする。(今できないけど)

実際に走らせてみて、「同じpwm値を指定しているのに、斜めに走る」って問題が出てからSpeedCalculatorを拡張する方がいいと思う。

**Comment by KakinokiKanta** [comment]
buildができないってのは、自分のPCのローカルでってこと?
CIは通ってると思うんだけど

File: module/Motion/Straight.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+Straight::Straight(int _targetValue, double _targetSpeed)
+  : targetValue(_targetValue), targetSpeed(_targetSpeed)
+{
+}
+
+void Straight::run()
+{
+  // 事前条件判定が真でないときは終了する
+  if(runPreconditionJudgement() != true) {
```
Comment (miyashita64) [inline module/Motion/Straight.cpp]: 好みの話
```suggestion
  if(!runPreconditionJudgement()) {
```

File: module/Motion/Straight.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    return;
+  }
+
+  // 直進前の走行距離
+  int initialRightMotorCount = Measurer::getRightCount();
+  int initialLeftMotorCount = Measurer::getLeftCount();
```
Comment (miyashita64) [inline module/Motion/Straight.cpp]: メンバで持っていいと思う。

File: module/Motion/Straight.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  // 走行距離が目標値に到達するまで繰り返す
+  while(true) {
+    // 現在の距離を取得する
+    int currentRightMotorCount = Measurer::getRightCount();
+    int currentLeftMotorCount = Measurer::getLeftCount();
+
+    // 事後条件が満たされたときループから抜ける オーバーライド必須
+    if(runPostconditionJudgement(initialRightMotorCount, initialLeftMotorCount,
+                                 currentRightMotorCount, currentLeftMotorCount)
```
Comment (miyashita64) [inline module/Motion/Straight.cpp]: 別コメントで挙げた
①左右の誤差はSpeedCalculatorが担うべき
②init〇〇Countをメンバする
を適用するなら、

current〇〇Countは、事後条件判定の中で取得してもいいと思う。色指定の場合はMotorCount必要ないし。

File: module/Motion/Straight.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    Controller::setRightMotorPwm(currentPwm - adjustmentPwm);
+
+    // 10ミリ秒待機
+    Timer.sleep(10);
+  }
+  // モータの停止
+  Controller::stopMotor();
+}
+
+bool Straight::runPreconditionJudgement()
```
Comment (miyashita64) [inline module/Motion/Straight.cpp]: オーバーライド前提なら、return false;だけ書いて、オーバライドできていない異常事態でも即座に動作を終了する方がいいと思った。

File: module/Motion/Straight.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#include "SpeedCalculator.h"
+#include "Timer.h"
+#include "SystemInfo.h"
+
+class Straight : public Motion {
+ public:
+  /**
+   * コンストラクタ　オーバーライド必須
+   * @param _targetValue 目標値 目標距離か目標色　オーバーライド必須
+   * @param _speed       目標速度[mm/s]
```
Comment (miyashita64) [inline module/Motion/Straight.h]:
```suggestion
   * @breif コンストラクタ
   * @param _targetValue 目標値 目標距離か目標色　オーバーライド必須
   * @param _speed       目標速度[mm/s]
   * @note オーバーライド必須
```
かな

他も同様

File: module/Motion/Straight.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+   * @brief 直進する
+   */
+  void run();
+
+  /**
+   * オーバーライド必須
+   * @brief 直進する際の事前条件判定をする
+   */
+  virtual bool runPreconditionJudgement();
```
Comment (miyashita64) [inline module/Motion/Straight.h]: is〇〇の方がbool値を返すことが明示的だと思う

**Review by miyashita64 — CHANGES_REQUESTED** [review]
無理に揃える必要はないけど、できるだけ回頭と近い形にしてほしい

File: module/Motion/Straight.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+   * @param currentLeftMotorCount  現在右モータ角位置
+   */
+  virtual bool runPostconditionJudgement(int initialRightMotorCount, int initialLeftMotorCount,
+                                         int currentRightMotorCount, int currentLeftMotorCount)
+      = 0;
+
+  /**
+   * @brief 実行のログを取る オーバーライド必須
+   */
+  virtual void logRunning();
```
Comment (bizyutyu) [inline module/Motion/Straight.h]: レビュー遅くなりました。
今気づいてる箇所だけ。修正お願いします。
```suggestion
  /**
   * @brief 実行のログを取る
   */
  void logRunning();
```

**Review by bizyutyu — CHANGES_REQUESTED** [review]

File: module/Motion/Straight.cpp (hunk 5 above)
Comment (miyashita64) [reply module/Motion/Straight.cpp]: そもそもオーバライド前提の関数は基底クラスのcppで実装しなくていいのかな？

File: module/Motion/Straight.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+bool Straight::isRunPreconditionJudgement()
+{
+  const int BUF_SIZE = 256;
+  char buf[BUF_SIZE];
+  // \"target\"をオーバーライド必須
+  // 目標速度値が0の場合は終了する
+  if(targetSpeed == 0) {
+    logger.logWarning("The speed value passed to \"target\"Straight is 0");
```
Comment (miyashita64) [inline module/Motion/Straight.cpp:58]:
```suggestion
    logger.logWarning("The speed value passed to \"target\" Straight is 0");
```

File: module/Motion/Straight.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    return false;
+  }
+  // \"target\"をオーバーライド必須
+  // pwmの絶対値がMIN_PWMより小さい場合はwarningを出す
+  SpeedCalculator SpeedCalculator(targetSpeed);
+  int pwm = SpeedCalculator.calcPwmFromSpeed();
+  if(abs(pwm) < MIN_PWM) {
+    snprintf(buf, BUF_SIZE, "The pwm value passed to \"target\"Straight is %d", pwm);
```
Comment (miyashita64) [inline module/Motion/Straight.cpp:66]:
```suggestion
    snprintf(buf, BUF_SIZE, "The pwm value passed to \"target\" Straight is %d", pwm);
```

File: module/Motion/Straight.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  /**
+   * @brief 直進する
+   */
+  void run();
+
+  /**
+   * @brief 直進する際の事前条件判定をする
+   * @note オーバーライド必須
+   */
+  virtual bool isRunPreconditionJudgement();
```
Comment (miyashita64) [inline module/Motion/Straight.h:35]:
```suggestion
  virtual bool isMetPrecondition();
```
とかのがいいかな？

File: module/Motion/Straight.cpp [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  return true;
+}
+
+void Straight::logRunning()
+{
+  const int BUF_SIZE = 128;
+  char buf[BUF_SIZE];  // log用にメッセージを一時保持する領域
+
+  // targetValueと%~のオーバーライド必須
+  snprintf(buf, BUF_SIZE, "Run \"targetValue\"Straight (\"targetValue\": %d, targetSpeed: %f)",
```
Comment (miyashita64) [inline module/Motion/Straight.cpp:85]:
```suggestion
  snprintf(buf, BUF_SIZE, "Run \"targetValue\" Straight (\"targetValue\": %d, targetSpeed: %f)",
```

**Review by KakinokiKanta — APPROVED** [review]
a

**Review by miyashita64 — APPROVED** [review]
LGTM!!
LGTM

**Review by aridome222 — APPROVED** [review]
LGTM!
LGTM

**Review by molpui0726 — APPROVED** [review]
LGTM!　お疲れ様でした🙇

**Review by bizyutyu — APPROVED** [review]
LGTM

**Pull request merged**